fmt/src/selection.rs

```diff
@@ -182,15 +180,45 @@ fn select_files_with_git(
         builder.build().context(SelectWithIgnoreSnafu)?
     };
 
-    let mut it = WalkDir::new(basedir).into_iter();
+    let basedir = basedir
+        .canonicalize()
+        .with_context(|_| ResolveAbsolutePathSnafu {
+            path: basedir.display().to_string(),
+        })?;
+    let mut it = WalkDir::new(basedir).follow_links(false).into_iter();
+
+    let workdir = repo
+        .work_dir()
+        .context(GixPathNotFountSnafu { path: "workdir" })?;
+    let workdir = workdir
+        .canonicalize()
+        .with_context(|_| ResolveAbsolutePathSnafu {
+            path: workdir.display().to_string(),
+        })?;
+    let worktree = repo
+        .worktree()
+        .context(GixPathNotFountSnafu { path: "worktree" })?;
+    let mut excludes = worktree
+        .excludes(None)
+        .map_err(Box::new)
+        .context(GixExcludeOpSnafu)?;
+
     while let Some(entry) = it.next() {
         let entry = entry.context(TraverseDirSnafu)?;
         let path = entry.path();
         let file_type = entry.file_type();
+
+        let rela_path = path
+            .strip_prefix(&workdir)
+            .expect("git repository encloses iteration");
+        let platform = excludes
+            .at_path(rela_path, Some(file_type.is_dir()))
+            .context(GixCheckExcludeOpSnafu)?;
+
         if file_type.is_symlink() {
```

Comment:
Since we skip symlink, perhaps we move the construction of `platform` down and make `!file_type.is_symlink()` a guard condition?

```rust
if file_type.is_symlink() { ... continue; }

let relpath = ...
let platform = ...
```

This is the only merge blocker I'd like to ask.

Reply:
```diff
diff --git a/fmt/src/selection.rs b/fmt/src/selection.rs
index 03045cb..5969e44 100644
--- a/fmt/src/selection.rs
+++ b/fmt/src/selection.rs
@@ -203,6 +203,10 @@ fn select_files_with_git(
         let entry = entry.context(TraverseDirSnafu)?;
         let path = entry.path();
         let file_type = entry.file_type();
+        if !file_type.is_file() && !file_type.is_dir() {
+            debug!(?file_type, "skip file: {path:?}");
+            continue;
+        }
 
         let rela_path = path
             .strip_prefix(&workdir)
@@ -210,10 +214,7 @@ fn select_files_with_git(
         let platform = excludes
             .at_path(rela_path, Some(file_type.is_dir()))
             .context(GixCheckExcludeOpSnafu)?;
-
-        if file_type.is_symlink() {
-            debug!("skip symlink: {:?}", path);
-        } else if file_type.is_dir() {
+        if file_type.is_dir() {
             if platform.is_excluded() {
                 debug!("skip git ignored directory: {:?}", path);
                 it.skip_current_dir();
@@ -234,8 +235,6 @@ fn select_files_with_git(
                 continue;
             }
             result.push(path.to_path_buf());
-        } else {
-            debug!("skip unknown file type ({:?}): {:?}", file_type, path);
         }
     }
```

Reply:
This may not help in performance a lot, since few files are not dir or regular files. But it's logically clear.

Reply:
True, a great catch, the code will be clearer that way without changing anything else.
Please feel free to apply the patch, it looks good to me.